server/service/src/programs/program_event.rs

```diff
@@ -873,4 +881,148 @@ mod test {
             ]
         )
     }
+
+    fn check_integrity(mut events: Vec<ProgramEventRow>) {
+        events.sort_by(|a, b| {
+            a.datetime
+                .cmp(&b.datetime)
+                .then_with(|| a.active_start_datetime.cmp(&b.active_start_datetime))
+        });
+        if events.is_empty() {
+            return;
+        }
+
+        // init with first datetime
+        let mut prev_event_end = events[0].datetime;
+        while !events.is_empty() {
+            let cur_stack_time = events[0].datetime;
+            // remove first stack, i.e., events with the same datetime
+            let mut stack = vec![];
+            while !events.is_empty() && events[0].datetime == cur_stack_time {
+                let e = events.remove(0);
+                stack.push(e);
+            }
+
+            // validate stack integrity
+            let stack_end = stack.last().unwrap().active_end_datetime;
+            let mut stack_end_time_reached = false;
+            for (i, event) in stack.iter().enumerate() {
+                assert!(event.datetime <= event.active_start_datetime);
+                assert!(event.datetime <= event.active_end_datetime);
+
+                if !stack_end_time_reached && event.active_end_datetime == stack_end {
+                    stack_end_time_reached = true;
+                }
+                if i == 0 {
+                    assert_eq!(
+                        event.datetime, prev_event_end,
+                        "Event {:?} must start where previous event ended",
+                        event.data
+                    );
+                } else if !stack_end_time_reached {
+                    assert_eq!(
+                        event.active_start_datetime, prev_event_end,
+                        "Event {:?} must start where previous event ended",
```

Comment:
"Stack end time not reached"?

Reply:
This is for within a stack. It verifies that there is only one event active at a time...